tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -477,201 +479,178 @@ func (a *AwsProvider) getInfrastructureVPC(infraID string) (*ec2.Vpc, error) {
+				Values: aws.StringSlice([]string{sgName}),
+			},
 		},
 	})
-	if err != nil {
-		return err
+	if err != nil || len(sgs.SecurityGroups) == 0 {
+		return nil, fmt.Errorf("worker security group not found, %v", err)
 	}
-	return nil
+
+	return sgs.SecurityGroups[0], nil
 }
 
-// findWindowsWorkerSg creates the Windows worker security group with name <infraID>-windows-worker-sg.
-func (a *AwsProvider) createWindowsWorkerSg(infraID string, vpc *ec2.Vpc) (string, error) {
+// createWindowsWorkerSg creates the Windows worker security group with name <infraID>-windows-worker-sg.
+func (a *AwsProvider) createWindowsWorkerSg(infraID string, vpc *ec2.Vpc) (*ec2.CreateSecurityGroupOutput, error) {
 	sgName := strings.Join([]string{infraID, "windows", "worker", "sg"}, "-")
 	sg, err := a.EC2.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
 		GroupName:   aws.String(sgName),
 		Description: aws.String("security group for RDP, winrm, ssh and all traffic within VPC"),
 		VpcId:       vpc.VpcId,
 	})
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	err = a.authorizeSgIngress(*sg.GroupId, vpc)
-	if err != nil {
-		return "", err
-	}
+	return sg, nil
+}
 
-	err = resource.AppendInstallerInfo([]string{}, []string{*sg.GroupId}, a.resourceTrackerDir)
-	if err != nil {
-		return "", fmt.Errorf("failed to record security group ID to file at '%s',"+
-			"security group will not be deleted, %v", a.resourceTrackerDir, err)
-	}
+// verifyAndUpdateSg verifies if an update is required to the rules in existing sg, or adding all the rules if the sg
+// is new. These rules are updated in the sg using addIngressRules. Returns error if rules
+// to be updated cannot be returned or  cannot be updated.
+func (a *AwsProvider) verifyAndUpdateSg(myIP string, sg *ec2.SecurityGroup, vpc *ec2.Vpc) ([]*ec2.IpPermission, error) {
 
-	return *sg.GroupId, nil
-}
+	// Get the list of rules to be updated, returns empty list is there are no rules to be updated
+	rules := a.GetRulesForSgUpdate(myIP, sg.IpPermissions, *vpc.CidrBlock)
 
-// authorizeSgIngress attaches inbound rules for RDP from user's external IP address and all traffic within the VPC
-// on newly created security group for Windows instance. The function returns error if failed.
-func (a *AwsProvider) authorizeSgIngress(sgID string, vpc *ec2.Vpc) error {
-	myIP, err := GetMyIp()
-	if err != nil {
-		return err
+	// call addIngressRules() only if there are any rules returned from GetRulesForSgUpdate()
+	if len(rules) != 0 {
+		err := a.addIngressRules(*sg.GroupId, rules)
+		if err != nil {
+			return nil, fmt.Errorf("could not set rules in sg %s: %s", *sg.GroupId, err)
+		}
 	}
+	return rules, nil
+}
 
-	_, err = a.EC2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
-		GroupId: aws.String(sgID),
-		IpPermissions: []*ec2.IpPermission{
+// GetRulesForSgUpdate returns a list of rules which are required to be updated in sg with local IP
+// This serves as an input to addIngressRules function.
```

Comment:
You can mention that the function guarantees that at least an empty slice is returned.